Ensure shared modules appear in params only once.

avital opened the pull request:
Concretely, create a shared submodule, then pass it in to a child module
and apply it there. Test that the parameters are only stored once, in
the frame where the shared module was created.

Co-authored-by: jheek <heekjonathan@gmail.com>

jheek commented on .gitignore:
```diff
@@ -4,3 +4,4 @@
 .DS_Store
 docs/_autosummary
 docs/_build
+.vscode
```
this should probably be removed

avital replied:
Done

avital commented:
Sorry for the late delay. I rewrote the docstring per the feedback you shared with me off-thread, and added a docstring for ModuleFrame to help explain the machinery.

Let me know if this looks good.

avital commented:
I merged with a rebase. Per your suggestion @jheek let's start doing that rather than merge commits unless it's genuinely a long merge branch.
